### PR TITLE
Request entity data

### PR DESCRIPTION
**Pull request by julien-me**
Request entity data from the entity server.

- [x] Add an `EntityQuery` class with a `write()` method.
- [x] Add an `EntityServer` class with `update()` and `queryOctree()` methods. 
- [x] Add an `OctreeQuery` class and implement it just enough to satisfy the needs of the above.
- [x] Add `packFloatAngleToTwoByte()` and `packClipValueToTwoByte()` to the class `GLMHelpers`.
- [x] Add a `ConicalViewFrustrum` class and implement it just enough to satisfy the needs of the above.
- [x] Update the example app with a game loop for the entity server.
- [x] Confirm receipt at entity server through instrumentation.

### CODE REVIEW

File: tests/domain/networking/packets/EntityQuery.unit.test.js [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+        const data = octreeQuery.getBroadcastData();
+
+        /* eslint-disable @typescript-eslint/no-magic-numbers */
+        data.connectionID = 27713;
+        data.conicalViews.at(0).position = { x: -20.461111, y: -5.7803667, z: -6.8709172 };
+        data.conicalViews.at(0).direction = { x: 0.86810995, y: 0, z: 0.49637193 };
+        data.conicalViews.at(0).angle = 0.714619;
+        data.conicalViews.at(0).farClip = 16384;
+        data.conicalViews.at(0).radius = 3;
```
Comment (ctrlaltdavid) [inline tests/domain/networking/packets/EntityQuery.unit.test.js:37]: These values should really be set via an `octreeQuery.setConicalViews()` method.

File: tests/domain/shared/GLMHelpers.unit.test.js [hunk 2]
```diff
@@ -91,3 +91,44 @@ describe("Quat - unit tests", () => {
...
 
 });
+
+describe("Angle - unit tests", () => {
+
+    test("Can write an angle into 2 bytes of packet data", () => {
+        const buffer = new ArrayBuffer(4);
+        const data = new DataView(buffer);
+
+        const angle = 45.2;
```
Comment (ctrlaltdavid) [inline tests/domain/shared/GLMHelpers.unit.test.js:101]: Can angles also be negative? If so also test a negative value.

**Review by ctrlaltdavid — CHANGES_REQUESTED** [review]

File: src/domain/networking/packets/EntityQuery.ts [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+        data.setInt32(dataPosition, info.maxQueryPPS, UDT.LITTLE_ENDIAN);
+        dataPosition += 4;
+        data.setFloat32(dataPosition, info.octreeElementSizeScale, UDT.LITTLE_ENDIAN);
+        dataPosition += 4;
+        data.setInt32(dataPosition, info.boundaryLevelAdjust, UDT.LITTLE_ENDIAN);
+        dataPosition += 4;
+        data.setUint16(dataPosition, Object.keys(info.jsonParameters).length, UDT.LITTLE_ENDIAN);
+        dataPosition += 2;
+
+        // WEBRTC TODO: Pack jsonParameters.
```
[the remark this answers is not in the document]
Comment (julien-me) [reply src/domain/networking/packets/EntityQuery.ts:93]: But in the C++ code bytes are written only when `_jsonParameters` isn't empty.
Here: https://github.com/vircadia/vircadia/blob/9f708e3c564e6aee41988778085863cd4c6fe185/libraries/octree/src/OctreeQuery.cpp#L79-L82

And I believe that `OctreeQuery::parseData(ReceivedMessage& message)` is the method used to unpack the data. If so then it tries to read the JSON parameters data only if `binaryParametersBytes` (length of `_jsonParameters`) is greater than 0.
Here: https://github.com/vircadia/vircadia/blob/9f708e3c564e6aee41988778085863cd4c6fe185/libraries/octree/src/OctreeQuery.cpp#L150-L161

If what you say is true, that JSON parameters are used only by the avatar mixer and server-side scripts, then I guess that we can simply ignore the JSON parameters in `EntityQuery.write()` and remove this TODO? :thinking:

[the remark this answers is not in the document]
Comment (ctrlaltdavid) [reply src/domain/networking/packets/EntityQuery.ts:93]: Yes, that's what I'm thinking: remove the TODO. But also add an `assert()` that the JSON parameters object is empty  / number of keys is 0. And in the JSDoc for `EntityQueryDetails` included a comment, "not used client-side".

[the remark this answers is not in the document]
Comment (julien-me) [reply src/domain/networking/packets/EntityQuery.ts]: :+1:

File: src/domain/networking/packets/EntityQuery.ts [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+import ConicalViewFrustum from "../../shared/ConicalViewFrustum";
+import GLMHelpers from "../../shared/GLMHelpers";
+import PacketTypeValue from "../udt/PacketHeaders";
+import UDT from "../udt/UDT";
+import NLPacket from "../NLPacket";
+
+
+type EntityQueryDetails = {
+    connectionID: number,
+    numFrustums: number,
```
Comment (ctrlaltdavid) [inline src/domain/networking/packets/EntityQuery.ts:22]: This property isn't necessary: the number is the same as the length of the array of frustums.

File: src/domain/octree/OctreeConstants.ts [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    /* eslint-disable class-methods-use-this */
+
+    get DEFAULT_MAX_OCTREE_PPS(): number {
+        // eslint-disable-next-line @typescript-eslint/no-magic-numbers
+        return 600;
+    }
+
+    get TREE_SCALE(): number {
+        // eslint-disable-next-line @typescript-eslint/no-magic-numbers
+        return 32768; // ~20 miles..
```
Comment (ctrlaltdavid) [inline src/domain/octree/OctreeConstants.ts:39]:
```suggestion
        return 32768; // ~20 miles.

```

File: src/domain/octree/OctreeQuery.ts [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+    NoFlags,
+    WantInitialCompletion
+}
+
+
+/*@devdoc
+ *  The <code>OctreeQuery</code> class
+ *  <p>C++: <code>class OctreeQuery : public NodeData</code></p>
+ *  @class OctreeQuery
+ *  @param {boolean} randomizeConnectionID - To assign, or not, a random number to the connectionID
```
Comment (ctrlaltdavid) [inline src/domain/octree/OctreeQuery.ts:34]:
```suggestion
 *  @param {boolean} randomizeConnectionID - <code>true</code> to use a random number for the initial connection ID, <code>false</code> to start at <code>0</code>.
```

File: src/domain/networking/packets/EntityQuery.ts [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+    jsonParameters: Record<string, unknown>,
+    queryFlags: OctreeQueryFlags,
+};
+
+
+const EntityQuery = new class {
+    // C++ N/A
+
+    /*@devdoc
+     *  Information returned by {@link PacketScribe|writing} a {@link PacketType(1)|EntityQuery} packet.
```
Comment (ctrlaltdavid) [inline src/domain/networking/packets/EntityQuery.ts:36]:
```suggestion
     *  Information needed for {@link PacketScribe|writing} a {@link PacketType(1)|EntityQuery} packet.

```

File: src/domain/networking/packets/EntityQuery.ts [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+     *  @typedef {object} PacketScribe.EntityQueryDetails
+     *  @property {number} connectionID - The octree query connection ID.
+     *  @property {number} numFrustums - The number of frustums.
+     *  @property {Array<ConicalViewFrustum>} conicalViews - An array of conical frustums.
```
Comment (ctrlaltdavid) [inline src/domain/networking/packets/EntityQuery.ts:40]:
```suggestion
     *  @property {Array<ConicalViewFrustum>} conicalViews - An array of conical view frustums.

```

**Review by ctrlaltdavid — CHANGES_REQUESTED** [review]
Also, JSDoc for EntityQuery packet needed in PacketHeaders.ts.

File: src/domain/octree/OctreeQuery.ts [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+     *  @param {boolean} reportInitialCompletion - The report initial completion flag.
+     */
+    setReportInitialCompletion(reportInitialCompletion: boolean): void {
+        // C++ void setReportInitialCompletion(bool reportInitialCompletion)
+        this.#_reportInitialCompletion = reportInitialCompletion;
+    }
+
+    /*@devdoc
+     *  Sets the array of conical frustums.
+     *  @param {Array<ConicalViewFrustum} views - An array of conical frustums.
```
Comment (ctrlaltdavid) [inline src/domain/octree/OctreeQuery.ts:95]:
```suggestion
     *  @param {Array<ConicalViewFrustum>} views - An array of conical frustums.

```

**Review by ctrlaltdavid — CHANGES_REQUESTED** [review]

**Review by ctrlaltdavid — APPROVED** [review]
